[JSC] Implment `@InternalPromise.@internalAll` in C++

sosukesuzuki opened the pull request:
#### a5f67b2bf6279b2a37c1c499fd82aa7bf0067398
<pre>
[JSC] Implment `@InternalPromise.@internalAll` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=303965">https://bugs.webkit.org/show_bug.cgi?id=303965</a>

Reviewed by Yusuke Suzuki.

This patch reverts 304888@main to reland 304331@main.

The original patch used PromiseAllResolveJob which calls resolve() to settle
the result promise. However, resolve() looks up the &quot;then&quot; property on the
resolution value, which can have user-observable side effects. For example,
if Object.prototype.then is set to a custom function, it would be called
when resolving the result array, breaking the InternalPromise pipeline.

InternalPromise requires using fulfill() instead of resolve() because
fulfill() directly settles the promise without looking up &quot;then&quot;. This is
the key difference between InternalPromise and regular Promise - InternalPromise
must not trigger any user-observable operations.

This patch:

- Adds InternalPromiseAllResolveJob which uses fulfill() instead of resolve()
- Updates JSInternalPromiseConstructor to use InternalPromiseAllResolveJob
- Also uses fulfill() for the empty array case
- Adds a test to verify Object.prototype.then is not called through module loader

---

This patch changes to implement `@InternalPromiseAll.@internalAll` in C++.

                                       TipOfTree                  Patched

internal-promise-internal-all        7.2714+-0.6089     ^      4.1862+-0.2744        ^ definitely 1.7370x faster

* JSTests/microbenchmarks/internal-promise-internal-all.js: Added.
* JSTests/modules/object-prototype-then-should-not-be-called-through-internal-promise.js: Added.
(Object.prototype.then):
* JSTests/modules/object-prototype-then-should-not-be-called-through-internal-promise/test.js: Added.
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/InternalPromiseConstructor.js: Removed.
* Source/JavaScriptCore/runtime/JSInternalPromiseConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::internalPromiseAllResolveJob):
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/Microtask.h:

Canonical link: <a href="https://commits.webkit.org/304930@main">https://commits.webkit.org/304930@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43932e0751973ca190cfb6397c5c6c143cfb7ed2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136894 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9254 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48181 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144630 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89867 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9957 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9100 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104678 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7286 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122653 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85516 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6923 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4620 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5218 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128855 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116254 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40835 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147387 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135380 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8937 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41405 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113035 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8955 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7508 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113365 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28803 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6847 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118935 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63134 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8985 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36989 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/168161 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8706 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72551 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43876 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8926 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8777 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/JavaScriptCore/runtime/JSInternalPromiseConstructor.cpp:
```diff
@@ -64,4 +75,65 @@ JSInternalPromiseConstructor::JSInternalPromiseConstructor(VM& vm, FunctionExecu
+    }
+
+    JSPromiseCombinatorsGlobalContext* globalContext = JSPromiseCombinatorsGlobalContext::create(vm, promise, values, jsNumber(length));
+    for (unsigned index = 0; index < length; ++index) {
```
Can this use fastForEach or one of the helpers that use `getDirectIndex` / `getIndexQuickly` when the array supports it?

Jarred-Sumner commented on Source/JavaScriptCore/runtime/JSInternalPromiseConstructor.cpp:
```diff
@@ -64,4 +75,65 @@ JSInternalPromiseConstructor::JSInternalPromiseConstructor(VM& vm, FunctionExecu
+        return JSValue::encode(promise);
+    }
+
+    JSArray* values = JSArray::tryCreate(vm, globalObject->arrayStructureForIndexingTypeDuringAllocation(ArrayWithContiguous), length);
```
Since `performPromiseThenWithInternalMicrotask` could take a bit, I wonder if it should use `MarkedArgumentBuffer` instead and then only create the array at the very end once all operations complete?

Jarred-Sumner commented on Source/JavaScriptCore/runtime/JSInternalPromiseConstructor.cpp:
```diff
@@ -64,4 +75,65 @@ JSInternalPromiseConstructor::JSInternalPromiseConstructor(VM& vm, FunctionExecu
+        return JSValue::encode(promise);
+    }
+
+    JSPromiseCombinatorsGlobalContext* globalContext = JSPromiseCombinatorsGlobalContext::create(vm, promise, values, jsNumber(length));
```
ah nvm since it needs the JSArray* the above comment wouldn't work

sosukesuzuki committed:
[JSC] Implment `@InternalPromise.@internalAll` in C++
https://bugs.webkit.org/show_bug.cgi?id=303965

Reviewed by Yusuke Suzuki.

This patch reverts 304888@main to reland 304331@main.

The original patch used PromiseAllResolveJob which calls resolve() to settle
the result promise. However, resolve() looks up the "then" property on the
resolution value, which can have user-observable side effects. For example,
if Object.prototype.then is set to a custom function, it would be called
when resolving the result array, breaking the InternalPromise pipeline.

InternalPromise requires using fulfill() instead of resolve() because
fulfill() directly settles the promise without looking up "then". This is
the key difference between InternalPromise and regular Promise - InternalPromise
must not trigger any user-observable operations.

This patch:

- Adds InternalPromiseAllResolveJob which uses fulfill() instead of resolve()
- Updates JSInternalPromiseConstructor to use InternalPromiseAllResolveJob
- Also uses fulfill() for the empty array case
- Adds a test to verify Object.prototype.then is not called through module loader

---

This patch changes to implement `@InternalPromiseAll.@internalAll` in C++.

                                       TipOfTree                  Patched

internal-promise-internal-all        7.2714+-0.6089     ^      4.1862+-0.2744        ^ definitely 1.7370x faster

* JSTests/microbenchmarks/internal-promise-internal-all.js: Added.
* JSTests/modules/object-prototype-then-should-not-be-called-through-internal-promise.js: Added.
(Object.prototype.then):
* JSTests/modules/object-prototype-then-should-not-be-called-through-internal-promise/test.js: Added.
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/InternalPromiseConstructor.js: Removed.
* Source/JavaScriptCore/runtime/JSInternalPromiseConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::internalPromiseAllResolveJob):
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/Microtask.h:

Canonical link: https://commits.webkit.org/304930@main
